typescript/sdk/src/core/app.ts

```diff
@@ -1,37 +1,55 @@
-import { Inbox, InboxValidatorManager } from '@abacus-network/core';
-
 import { AbacusApp } from '../app';
-import { domains } from '../domains';
-import { ChainName, NameOrDomain } from '../types';
+import { MultiProvider } from '../provider';
+import { ChainName, Remotes } from '../types';
+import { objMap } from '../utils';
+import {
+  CoreContractAddresses,
+  CoreContracts,
+  CoreContractSchema,
+} from './contracts';
+import { environments } from './environments';
 
-import { CoreContractAddresses, CoreContracts } from './contracts';
+type Environments = typeof environments;
+type EnvironmentName = keyof Environments;
 
-export class AbacusCore extends AbacusApp<
-  CoreContractAddresses,
-  CoreContracts
-> {
-  constructor(addresses: Partial<Record<ChainName, CoreContractAddresses>>) {
-    super();
-    const chains = Object.keys(addresses) as ChainName[];
-    chains.map((chain) => {
-      this.registerDomain(domains[chain]);
-      const domain = this.resolveDomain(chain);
-      this.contracts.set(domain, new CoreContracts(addresses[chain]!));
-    });
+export class AbacusCore<
+  Networks extends ChainName = ChainName,
+> extends AbacusApp<CoreContracts<Networks>, Networks> {
+  constructor(
+    networkAddresses: {
+      [local in Networks]: CoreContractAddresses<Networks, local>;
+    },
+    multiProvider: MultiProvider<Networks>,
+  ) {
+    super(
+      objMap<Networks, any, any>(networkAddresses, (local, addresses) => {
+        return new CoreContracts<Networks, typeof local>(
+          addresses,
+          multiProvider.getDomainConnection(local).getConnection()!,
+        );
+      }),
+    );
+  }
+
+  static fromEnvironment(
+    name: EnvironmentName,
+    multiProvider: MultiProvider<keyof Environments[typeof name]>,
+  ) {
+    return new AbacusCore(environments[name], multiProvider);
   }
 
-  mustGetInbox(src: NameOrDomain, dest: NameOrDomain): Inbox {
-    const contracts = this.mustGetContracts(dest);
-    const srcName = this.mustGetDomain(src).name;
-    return contracts.inbox(srcName);
+  getContracts<Local extends Networks>(
+    network: Local,
+  ): CoreContractSchema<Networks, Local> {
+    return this.get(network).contracts as any;
```

Comment:
Why do we need to cast to here?

Reply:
the `Remotes<Networks, Local>` thing specific to `core`